### PR TITLE
feat(api): add kafka platform for hook provider

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/platform/dao_model.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+func LoadModel(db gorp.SqlExecutor, modelID int64) (sdk.PlatformModel, error) {
+	var pm PlatformModel
+	if err := db.SelectOne(&pm, "SELECT * from platform_model where id = $1", modelID); err != nil {
+		return sdk.PlatformModel{}, sdk.WrapError(err, "LoadModel> Cannot select platform model %d", modelID)
+	}
+	return sdk.PlatformModel(pm), nil
+}
+
+// CreatePlatformModels create platforms models
+func CreatePlatformModels(db *gorp.DbMap) error {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: CreateModels

Comment (fsamin) [inline engine/api/platform/dao_model.go]: what about using a repository pattern like in the services package (just asking), dunno it's this king of pattern is really a good approach

File: engine/api/platform/dao_model.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	"encoding/json"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+)
+
+var (
+	Models = []sdk.PlatformModel{
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: missing comment

File: engine/api/platform.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package api
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/ovh/cds/engine/api/platform"
+)
+
+func (api *API) getPlatformModel() Handler {
```
Comment (fsamin) [inline engine/api/platform.go]: models

File: engine/api/platform.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		return WriteJSON(w, platform.Models, http.StatusOK)
```
Comment (fsamin) [inline engine/api/platform.go]: shouldn't we load the models from the database ?

File: engine/api/platform/dao_model.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+			// update default values
+			if err := UpdatePlatformModel(tx, p); err != nil {
+				return sdk.WrapError(err, "CreatePlatformModels  error on update")
+			}
+		}
+	}
+	return tx.Commit()
+}
+
+func checkPlatformExist(db gorp.SqlExecutor, p *sdk.PlatformModel) (bool, error) {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: ModelExists(db, modelName)

File: engine/sql/082_platform.sql [hunk 6]
```diff
@@ -0,0 +1,29 @@
+-- +migrate Up
+CREATE TABLE platform_model (
+  id BIGSERIAL PRIMARY KEY,
+  name VARCHAR(50),
```
Comment (fsamin) [inline engine/sql/082_platform.sql]: isn't the name unique ?

File: engine/sql/082_platform.sql [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  file_storage BOOLEAN,
+  block_storage BOOLEAN,
+  deployment BOOLEAN,
+  compute BOOLEAN
+);
+
+CREATE TABLE project_platform (
+  id BIGSERIAL PRIMARY KEY,
+  name VARCHAR(100),
+  project_id BIGINT,
```
Comment (fsamin) [inline engine/sql/082_platform.sql]: no foreign key constraint with project ?

File: engine/api/platform/dao_model.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	var count = 0
+	if err := db.QueryRow("select count(1), id from platform_model where name = $1 group by id", p.Name).Scan(&count, &p.ID); err != nil {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: is name unique ? why grouping by id ?

File: engine/api/platform/dao_model.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+		if err == sql.ErrNoRows {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: a select count can't return no rows

File: engine/api/platform/dao_model.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+			return false, nil
+		}
+		return false, sdk.WrapError(err, "checkPlatformExist")
+	}
+	return count > 0, nil
+}
+
+// InsertPlatformModel inserts a platform model in database
+func InsertPlatformModel(db gorp.SqlExecutor, m *sdk.PlatformModel) error {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: InsertModel

File: engine/api/platform/dao_model.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	dbm := PlatformModel(*m)
+	if err := db.Insert(&dbm); err != nil {
+		return sdk.WrapError(err, "InsertPlatformModel> Unable to insert platform model %s", m.Name)
+	}
+	*m = sdk.PlatformModel(dbm)
+	return nil
+}
+
+// UpdatePlatformModel updates a platform model in database
+func UpdatePlatformModel(db gorp.SqlExecutor, m *sdk.PlatformModel) error {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: UpdateModel

File: engine/api/platform/dao_model.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	if n, err := db.Update(&dbm); err != nil {
+		return sdk.WrapError(err, "UpdatePlatformModel> Unable to update platform model %s", m.Name)
+	} else if n == 0 {
+		return sdk.WrapError(sdk.ErrNotFound, "UpdatePlatformModel> Unable to update platform model %s", m.Name)
+	}
+	return nil
+}
+
+// PostGet is a db hook
+func (pm *PlatformModel) PostGet(db gorp.SqlExecutor) error {
```
Comment (fsamin) [inline engine/api/platform/dao_model.go]: i think the type PlatformModel could be private

File: engine/api/platform/gorp_model.go [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package platform
+
+import (
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+
+	"github.com/ovh/cds/sdk"
+)
+
+// PlatformModel is a gorp wrapper around sdk.PlatformModel
+type PlatformModel sdk.PlatformModel
```
Comment (fsamin) [inline engine/api/platform/gorp_model.go]: this type could be private

File: engine/api/project/dao_platform.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+		return sdk.WrapError(err, "dbProjectPlatform.PostGet> Cannot load model")
+	}
+	pp.Model = model
+
+	query := "SELECT config FROM project_platform where id = $1"
+	s, err := db.SelectNullStr(query, pp.ID)
+	if err != nil {
+		return sdk.WrapError(err, "dbProjectPlatform.PostGet> Cannot get config")
+	}
+	if s.Valid {
```
Comment (fsamin) [inline engine/api/project/dao_platform.go]: why not using `gorpmapping.JSONNullString`

File: engine/api/project/dao_platform.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+		if err := json.Unmarshal([]byte(s.String), &config); err != nil {
+			return sdk.WrapError(err, "dbProjectPlatform.PostGet> Cannot unmarshall config")
+		}
+		pp.Config = config
+	}
+	return nil
+}
+
+// LoadPlatformsByName Load a platform by project key and its name
+func LoadPlatformsByName(db gorp.SqlExecutor, key string, name string, clearPwd bool) (sdk.ProjectPlatform, error) {
```
Comment (fsamin) [inline engine/api/project/dao_platform.go]: LoadByName

File: engine/api/project/dao_platform.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+				p.Config[k] = v
+			}
+		}
+	}
+
+	return p, nil
+}
+
+// LoadPlatformsByID load project platforms by project id
+func LoadPlatformsByID(db gorp.SqlExecutor, id int64, clearPassword bool) ([]sdk.ProjectPlatform, error) {
```
Comment (fsamin) [inline engine/api/project/dao_platform.go]: LoadByProjectID

File: engine/api/project/dao_platform.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+func encryptPlatformValue(v string) (string, error) {
+	encryptedSecret, errE := secret.Encrypt([]byte(v))
+	if errE != nil {
+		return "", sdk.WrapError(errE, "encryptPlatformValue> Cannot encrypt password")
+	}
+	return base64.StdEncoding.EncodeToString(encryptedSecret), nil
+}
+
+// InsertPlatform inserts a project platform
+func InsertPlatform(db gorp.SqlExecutor, pp *sdk.ProjectPlatform) error {
```
Comment (fsamin) [inline engine/api/project/dao_platform.go]: Insert

File: engine/api/project/dao_platform.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+	if err := db.Insert(&ppDb); err != nil {
+		return sdk.WrapError(err, "InsertPlatform> Cannot insert project platform")
+	}
+	*pp = sdk.ProjectPlatform(ppDb)
+	return nil
+}
+
+// PostInsert is a db hook
+func (pp *dbProjectPlatform) PostInsert(db gorp.SqlExecutor) error {
+	configB, err := json.Marshal(pp.Config)
```
Comment (fsamin) [inline engine/api/project/dao_platform.go]: why not use `gorpmapping.JSONToNullString`

File: engine/sql/082_platform.sql [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  disabled BOOLEAN,
+  hook BOOLEAN,
+  file_storage BOOLEAN,
+  block_storage BOOLEAN,
+  deployment BOOLEAN,
+  compute BOOLEAN
+);
+
+CREATE TABLE project_platform (
+  id BIGSERIAL PRIMARY KEY,
```
Comment (fsamin) [inline engine/sql/082_platform.sql]: why this association table have an id ? shouldn't the primary key composed by model_id and project_id ?

File: engine/api/project/gorp_model.go [hunk 20]
```diff
@@ -15,14 +15,15 @@ import (
...
-type dbVariable sdk.Variable
 type dbProjectVariableAudit sdk.ProjectVariableAudit
 type dbProjectKey sdk.ProjectKey
+type dbProjectPlatform sdk.ProjectPlatform
 
 func init() {
 	gorpmapping.Register(gorpmapping.New(dbProject{}, "project", true, "id"))
 	gorpmapping.Register(gorpmapping.New(dbProjectVariableAudit{}, "project_variable_audit", true, "id"))
 	gorpmapping.Register(gorpmapping.New(dbProjectKey{}, "project_key", false))
+	gorpmapping.Register(gorpmapping.New(dbProjectPlatform{}, "project_platform", true, "id"))
```
Comment (fsamin) [inline engine/api/project/gorp_model.go]: de we really need an autoincrement id for this ?

File: engine/api/project_platform.go [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+	"context"
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/ovh/cds/engine/api/project"
+	"github.com/ovh/cds/sdk"
+)
+
+func (api *API) getProjectPlatformHandler() Handler {
```
Comment (fsamin) [inline engine/api/project_platform.go]: unit tests ftw

File: ui/src/assets/i18n/fr.json [hunk 22]
```diff
@@ -471,6 +472,12 @@
   "project_variable_form_title": "Ajouter une nouvelle variable: ",
   "project_workflows_list": "Liste des workflows du projet: ",
 
+  "platform_name": "Nom de la plateforme",
+  "platform_model": "Model de la plateforme",
```
Comment (fsamin) [inline ui/src/assets/i18n/fr.json]: modèle

File: engine/api/project/dao_platform.go (hunk 15 above)
Comment (sguiheux) [reply engine/api/project/dao_platform.go]: no,  we are in package project

File: engine/api/project/dao_platform.go (hunk 16 above)
Comment (sguiheux) [reply engine/api/project/dao_platform.go]: no,  we are in package project

File: engine/api/project/dao_platform.go (hunk 17 above)
Comment (sguiheux) [reply engine/api/project/dao_platform.go]: no,  we are in package project

File: engine/api/project/dao_platform.go (hunk 14 above)
Comment (sguiheux) [reply engine/api/project/dao_platform.go]: don't want to create a type for 1 field :/

File: engine/sql/082_platform.sql (hunk 19 above)
Comment (sguiheux) [reply engine/sql/082_platform.sql]: you can add the same model any time you want,

File: engine/api/platform/dao_model.go (hunk 8 above)
Comment (sguiheux) [reply engine/api/platform/dao_model.go]: need to set id :)

File: engine/sql/082_platform.sql (hunk 6 above)
Comment (sguiheux) [reply engine/sql/082_platform.sql]: fixed

File: engine/sql/082_platform.sql (hunk 7 above)
Comment (sguiheux) [reply engine/sql/082_platform.sql]: fixed

File: engine/api/platform/dao_model.go [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+// LoadModel Load a platform model by its ID
+func LoadModel(db gorp.SqlExecutor, modelID int64) (sdk.PlatformModel, error) {
+	var pm platformModel
+	if err := db.SelectOne(&pm, "SELECT * from platform_model where id = $1", modelID); err != nil {
+		return sdk.PlatformModel{}, sdk.WrapError(err, "LoadModel> Cannot select platform model %d", modelID)
+	}
+	return sdk.PlatformModel(pm), nil
+}
+
+// CreatePlatformModels create platforms models
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: // CreateModels creates...

File: engine/api/platform/dao_model.go [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+			}
+		}
+	}
+	return tx.Commit()
+}
+
+func ModelExists(db gorp.SqlExecutor, p *sdk.PlatformModel) (bool, error) {
+	var count = 0
+	if err := db.QueryRow("select count(1), id from platform_model where name = $1 GROUP BY id", p.Name).Scan(&count, &p.ID); err != nil {
+		return false, sdk.WrapError(err, "checkPlatformExist")
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: "ModelExists>"

File: engine/api/platform/dao_model.go [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	return count > 0, nil
+}
+
+// InsertPlatformModel inserts a platform model in database
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: // InsertModel inserts...

File: engine/api/platform/dao_model.go [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+func InsertModel(db gorp.SqlExecutor, m *sdk.PlatformModel) error {
+	dbm := platformModel(*m)
+	if err := db.Insert(&dbm); err != nil {
+		return sdk.WrapError(err, "InsertPlatformModel> Unable to insert platform model %s", m.Name)
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: "InsertModel>"

File: engine/api/platform/dao_model.go [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	*m = sdk.PlatformModel(dbm)
+	return nil
+}
+
+// UpdatePlatformModel updates a platform model in database
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: // UpdateModel updates..

File: engine/api/platform/dao_model.go [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+func UpdateModel(db gorp.SqlExecutor, m *sdk.PlatformModel) error {
+	dbm := platformModel(*m)
+	if n, err := db.Update(&dbm); err != nil {
+		return sdk.WrapError(err, "UpdatePlatformModel> Unable to update platform model %s", m.Name)
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: "UpdateModel>"

File: engine/api/platform/dao_model.go [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+	} else if n == 0 {
+		return sdk.WrapError(sdk.ErrNotFound, "UpdatePlatformModel> Unable to update platform model %s", m.Name)
```
Comment (yesnault) [inline engine/api/platform/dao_model.go]: "UpdateModel>"

File: engine/api/project/dao_dependencies.go [hunk 30]
```diff
@@ -99,6 +99,24 @@ var (
...
+			return sdk.WrapError(err, "loadPlatforms> Cannot load platforms")
+		}
+		proj.Platforms = pf
+		return nil
+	}
+
+	loadClearPlatforms = func(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, u *sdk.User) error {
+		pf, err := LoadPlatformsByID(db, proj.ID, true)
+		if err != nil {
+			return sdk.WrapError(err, "loadPlatforms> Cannot load platforms")
```
Comment (yesnault) [inline engine/api/project/dao_dependencies.go]: "loadClearPlatforms>"

File: engine/hooks/kafka.go [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+	s.Dao.SaveTaskExecution(exec)
+}
+
+func (s *Service) startKafkaHook(t *sdk.Task) error {
+	var kafkaPlatform, kafkaUser, projectKey, consumerGroup, topic string
+	for k, v := range t.Config {
+		switch k {
+		case "platform":
+			kafkaPlatform = v.Value
+		case "consumer group":
```
Comment (yesnault) [inline engine/hooks/kafka.go]: const for theses kafka keys could be great "consumer group" / "platform" / "topic"

File: sdk/platform.go [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+type PlatformModel struct {
+	ID            int64          `json:"id" db:"id"`
+	Name          string         `json:"name" db:"name"`
+	Author        string         `json:"author" db:"author"`
+	Identifier    string         `json:"identifier" db:"identifier"`
+	Icon          string         `json:"icon" db:"icon"`
+	DefaultConfig PlatformConfig `json:"default_config" db:"-"`
+	Disabled      bool           `json:"disabled" db:"disabled"`
+	Hook          bool           `json:"hook" db:"hook"`
+	FileStorage   bool           `json:"fileStorage" db:"file_storage"`
```
Comment (yesnault) [inline sdk/platform.go]: json:"file_storage"

File: sdk/platform.go [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+	BlockStorage  bool           `json:"blockStorage" db:"block_storage"`
```
Comment (yesnault) [inline sdk/platform.go]: json:"block_storage"

File: sdk/project.go [hunk 34]
```diff
@@ -30,6 +30,7 @@ type Project struct {
 	WorkflowMigration string             `json:"workflow_migration" yaml:"workflow_migration" db:"workflow_migration"`
 	Keys              []ProjectKey       `json:"keys" yaml:"keys" db:"-" cli:"-"`
 	VCSServers        []ProjectVCSServer `json:"vcs_servers" yaml:"vcs_servers" db:"-" cli:"-"`
+	Platforms         []ProjectPlatform  `json:"platforms" yaml:"platform" db:"-" cli:"-"`
```
Comment (yesnault) [inline sdk/project.go]: yaml:"platforms"

File: ui/src/app/app.component.ts [hunk 35]
```diff
@@ -69,7 +69,7 @@ export class AppComponent  implements OnInit {
                 this.stopWorker(this.warningWorker, this.warningWorkerSubscription);
             } else {
                 this.isConnected = true;
-                this.startLastUpdateSSE();
+                // this.startLastUpdateSSE();
```
Comment (yesnault) [inline ui/src/app/app.component.ts]: uncomment

File: ui/src/assets/i18n/fr.json [hunk 36]
```diff
@@ -186,6 +186,7 @@
   "common_pipeline_title": "Pipeline : ",
   "common_pipeline": "Pipeline",
   "common_pipelines": "Pipelines",
+  "common_platforms": "Platformes",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json]: Plateformes

File: engine/api/project_platform.go (hunk 21 above)
Comment (sguiheux) [reply engine/api/project_platform.go]: done

File: sdk/cdsclient/client_project_kafka.go [hunk 37]
```diff
@@ -0,0 +1,14 @@
+package cdsclient
+
+import (
+	"fmt"
+	"github.com/ovh/cds/sdk"
```
Comment (bnjjj) [inline sdk/cdsclient/client_project_kafka.go]: blank line

File: ui/src/app/service/project/project.service.ts [hunk 38]
```diff
@@ -344,4 +345,43 @@ export class ProjectService {
...
+        return this._http.post<ProjectPlatform>('/project/' + key + '/platforms', p);
+    }
+
+    /**
+     * Remove a project platform
+     * @param key project unique key
+     * @param name platform name
+     * @returns {Observable<Object>}
+     */
+    removePlatform(key: string, name: string) {
```
Comment (bnjjj) [inline ui/src/app/service/project/project.service.ts]: return type --> Observable<any>

File: ui/src/app/views/project/show/platforms/form/platform.form.component.ts [hunk 39]
```diff
@@ -0,0 +1,41 @@
...
+        ProjectPlatform.mergeConfig(this.newPlatform.model.default_config, this.newPlatform.config);
+        console.log(this.newPlatform);
+    }
+
+    create(): void {
+        this.loading = true;
+        this._projectStore.addPlatform(this.project.key, this.newPlatform).pipe(first(), finalize(() => {
+            this.loading = false;
+        })).subscribe(() => {
+
```
Comment (bnjjj) [inline ui/src/app/views/project/show/platforms/form/platform.form.component.ts]: missing toast ?

File: ui/src/app/views/project/show/platforms/form/platform.form.component.ts [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+    constructor(private _platformService: PlatformService, private _projectStore: ProjectStore) {
+        this.newPlatform = new ProjectPlatform();
+        this._platformService.getPlatformModels().pipe(first()).subscribe(platfs => {
+            this.models = platfs;
+        });
+    }
+
+    updateConfig(): void {
+        ProjectPlatform.mergeConfig(this.newPlatform.model.default_config, this.newPlatform.config);
+        console.log(this.newPlatform);
```
Comment (bnjjj) [inline ui/src/app/views/project/show/platforms/form/platform.form.component.ts]: to delete

File: ui/src/assets/i18n/en.json [hunk 41]
```diff
@@ -471,6 +472,15 @@
   "project_variable_form_title": "Add a new variable: ",
   "project_workflows_list": "List of workflows in the project: ",
 
+  "platform_add_title" : "Link a platform: ",
+  "platform_list_title" : "Linked platform list: ",
+  "platform_name": "Platform name",
+  "platform_no" : "There is no liked platform",
```
Comment (bnjjj) [inline ui/src/assets/i18n/en.json]: linked

**Comment by bnjjj** [comment]
https://github.com/ovh/cds/issues/2247

**Pull request merged**